Run cargo build and read JSON output

ede1998 opened the pull request:
## About
This PR changes `cargo flamegraph` so that it reads the output of `cargo build` to determine the patch of the binary that should be profiled.
Because of this change the automatic target selection had to be redone as previously it just chose the only binary from the target folder. Now, we run `cargo metadata`, find the only binary target of the workspace first and pass this target explictly. If there are multiple or no binary targets, we abort with an error.
I tested the new version several different dummy crates and workspaces as well as `ripgrep`. It seems to work fine.

## Relevant issue
- #139 This is where the idea came up.
- #132  This should be fixed by this PR assuming `cargo build` tells us the correct binary.
- #79 This should also be fixed.

## Follow-up PR
- #142

djc commented:
Hey, this is looking great. I'm on vacation so won't be able to look at this for a while, but I'll make sure to get back to it in a few weeks.

ede1998 commented:
Thanks for the heads-up. Have a great vacation.

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -174,91 +176,86 @@ fn build(opt: &Opt) {
+}
 
-        let mut has_debuginfo = false;
+fn check_debug_info(opt: &Opt, artifacts: &[Artifact]) {
```
Let's not create a single-use function here, and keep the checking of debug info inside `build()` as before. In general, this kind of refactoring should also go in a separate commit to make it easier to review.

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -174,91 +176,86 @@ fn build(opt: &Opt) {
+        .expect("failed to execute cargo build command");
+
+    let messages = Message::parse_stream(&*stdout);
+    let artifacts: Vec<_> = messages
```
Is there a reason to bind `messages` here? It looks like we can call `filter_map()` directly on the return value of `Message:parse_stream()`.

ede1998 replied:
No. I think I originally did this because the code didn't look good after `cargo fmt` because of the 60 characters max width in `.rustfmt.toml`. But it seems like I changed the code again so it looks just fine now.

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -203,54 +211,30 @@ fn build(opt: &Opt) -> Vec<Artifact> {
+        eprintln!(
+            "cargo build failed: {}",
+            String::from_utf8_lossy(&stderr)
+        );
```
This should live in the previous commit.

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -203,54 +211,30 @@ fn build(opt: &Opt) -> Vec<Artifact> {
-                    },
-                );
-    }
+fn check_debug_info(opt: &Opt, artifact: &Artifact) {
+    const NONE: u32 = 0;
+    let debug = artifact.profile.debuginfo.unwrap_or(NONE);
```
These changes seem to be unrelated to the main goal of this commit?

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -259,109 +243,63 @@ fn check_debug_info(opt: &Opt, artifacts: &[Artifact]) {
-        .filter(|t| t.crate_types.contains(&"bin".into()))
-        .map(|t| t.name)
-        .collect();
+fn select_executable(
```
Prefer to inline this single-use function, as well.

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -306,11 +306,47 @@ fn workload(
+        .exec()
+        .expect("failed to access crate metadata");
+
+    let bin_targets: Vec<String> = metadata
```
Let's turn this into a single call chain?

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -306,11 +306,47 @@ fn workload(
+        .into_iter()
+        .flat_map(|p| p.targets.into_iter())
+        .filter(|t| t.kind.contains(&"bin".into()))
+        .map(|t| t.name)
```
We should use `filter_map()` instead of `filter().map()`. Ideally we should combine it with `flat_map()`, too, if at all possible.

ede1998 replied:
I combined it to `filter_map` but I didn't see a way to combine it with `flat_map`.

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -306,11 +306,47 @@ fn workload(
     result
 }
 
+fn find_unique_bin_target() -> String {
```
Let's also inline this single-use function.

ede1998 replied:
Sure I can inline them but it is quite long. That's why I split it off into another function to keep it more readable. Because of the function name, you can immediately see what it does which is not be the case if I'll inline the block (unless I add a describing comment).
The same goes for the other single use functions.

ede1998 replied:
I inlined all the other functions you suggested but I didn't inline this one yet. If you insist, I can inline it but I think `main` will be too long then.

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -153,7 +153,7 @@ enum Opts {
 
 fn build(opt: &Opt) -> Vec<Artifact> {
-    use std::process::{Command, Output};
+    use std::process::{Command, Output, Stdio};
```
Changes from this commit should be inlined into the previous commit.

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -258,6 +258,11 @@ fn check_debug_info(opt: &Opt, artifact: &Artifact) {
         );
         eprintln!("[profile.{}]", profile);
         eprintln!("debug = true\n");
+        eprintln!("Or set this environment variable:\n");
+        eprintln!(
+            "CARGO_PROFILE_{}_DEBUG=true\n",
+            profile.to_uppercase()
+        );
```
Inline changes from this commit into the relevant previous commits, too, please.

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -320,19 +322,46 @@ fn workload(
-        .filter(|t| t.kind.contains(&"bin".into()))
-        .map(|t| t.name)
+        .flat_map(|p| {
+            let Package { targets, name, .. } = p;
+            targets
```
Why can't we use `p.targets` directly anymore, as before?

ede1998 replied:
No because `p` is used in the closure again and the borrow checker does not like that:
```
error[E0382]: use of partially moved value: `p`
   --> src/bin/cargo-flamegraph.rs:356:22
    |
355 |                 .into_iter()
    |                  ----------- `p.targets` partially moved due to this method call
356 |                 .map(move |t| (p.name.clone(), t))
    |                      ^^^^^^^^  ------ use occurs due to use in closure
    |                      |
    |                      value used here after partial move
    |
note: this function takes ownership of the receiver `self`, which moves `p.targets`
```

djc reviewed (commented):
So here's an initial pass. As is, I found this pretty challenging to review, please follow up on these comments and I'll do another pass. (If you haven't adjusted commit history before, I usually use interactive rebases via `git rebase -i`, where you can use the `squash` and `fixup` commands to fold newer commits into older ones.)

ede1998 commented:
@djc Ok, I fixed up the commit history like you suggested and also implemented (most of) the other suggestions. I had to force-push to get the new history up here but I think that was to be expected. It seems though that the CI-workflow didn't automatically start again and I don't have permission to do that:
`First-time contributors need a maintainer to approve running workflows.`

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -174,28 +208,31 @@ fn build(opt: &Opt) {
-    let exit_status = child.wait().expect(
-        "failed to wait for cargo build child to finish",
-    );
+    let artifacts: Vec<_> = Message::parse_stream(&*stdout)
```
Ah, I missed that this now duplicates the `Message::parse_stream()` call. In that case, it still makes sense to factor it out so that `!opt.dev` can share the same result.

ede1998 replied:
I'll leave that as is because the second `Message::parse_stream` is removed in this PR anyway.

djc replied:
Ah, that's fine then.

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -174,28 +208,31 @@ fn build(opt: &Opt) {
-    );
+    let artifacts: Vec<_> = Message::parse_stream(&*stdout)
+        .filter_map(|m| match m {
+            Ok(Message::CompilerArtifact(artifact)) => {
```
This might be easier to folllow if you use `m.map()` and `collect()` into a `Result`, which you can then just expect on the outside -- maybe try it and see which one you think is easier to follow?

ede1998 replied:
I'd say it's a bit harder to read. But maybe that's just because I have to get used to it. While I like the `expect` on the outside, the `match` and `transpose` combo is a bit complex to read in my opinion. If you have any suggestions to simplify this, that would be great. The only way I see right now is to add an extension trait for `Message` that provides a `fn into_artifact(self) -> Option<Artifact>` but that seems like too much.

```rust
let artifacts = Message::parse_stream(&*stdout)
    .filter_map(|r| {
        r.map(|m| match m {
            Message::CompilerArtifact(a) => Some(a),
            _ => None,
        })
        .transpose()
    })
    .collect::<Result<Vec<_>, _>>()
    .expect("failed to parse cargo build output");
```

djc replied:
That's fair -- go with the approach you like best.

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -256,116 +293,59 @@ fn build(opt: &Opt) {
+                && a.executable.is_some()
+        });
+
+        let artifact = artifact.unwrap_or_else(|| {
```
Let's chain this directly? I think we can return `artifacts.iter().find(...).unwrap_or_else().executable.as_deref().expect()`? Also I'd probably leave that outer `expect()` to the caller.

ede1998 replied:
> I think we can return ...?

In this commit, yes but in the next commit 010ad1a8a6d3b04dce6dc49ada8e1641d669e282 I'm adding the debug info check inbetween `unwrap_or_else` and `executable`. This check needs the artifact.

> Also I'd probably leave that outer expect() to the caller.

It's not a function anymore just a block expression. Also the expect itself should never ever occur unless someone changes this part of the code again because the `find` predicate only allows artifacts with executables.
```rust
let binary_path = {
    let artifact = artifacts.iter().find(...).unwrap_or_else(...);
    artifact.executable.as_deref().expect(...)
};
```

djc replied:
Can we express that in some way that we don't have the unnecessary `expect()` call, then? I guess we could `find_map()` in a way that we return the deref of executable directly along with the `artifact`?

ede1998 replied:
Excellent suggestion. I like the new version much more.

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -337,6 +275,33 @@ fn workload(
+            artifact.profile.debuginfo.unwrap_or(NONE);
+
+        if !opt.dev && debug_level == NONE {
+            let profile = if opt.bench.is_some() {
+                "bench"
+            } else {
+                "release"
+            };
```
This would be more succinctly expressed as a plain `match` expression.

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -316,11 +316,53 @@ fn workload(
+                "automatically selected {} as it is the only binary target",
+                target
+            );
+            target.clone()
```
I think we can return `bin_targets[0]` here instead of cloning?

ede1998 replied:
No:
```
Error[E0507]: cannot move out of index of `Vec<BinaryTarget>`
   --> src/bin/cargo-flamegraph.rs:367:13
    |
367 |             bin_targets[0]
    |             ^^^^^^^^^^^^^^ move occurs because value has type `BinaryTarget`, which does not implement the `Copy` trait
```

I managed to make it work but I'm not convinced that's better:
```rust
match bin_targets.as_slice() {
    [_] => bin_targets.into_iter().next().unwrap(),
...
```

While developing I also tried to destructure the `Vec` itself but that also leads to borrow checker issues and unsized locals for the catch-all arm.

ede1998 replied:
Ok, just had another look into `Vec` docs. `bin_targets.remove(0);`works as well if I make `bin_targets` mutable. I think I will go with that.

djc replied:
Yup, that sounds good!

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -316,11 +316,53 @@ fn workload(
+    let bin_targets: Vec<String> = metadata
+        .packages
+        .into_iter()
+        .flat_map(|p| p.targets.into_iter())
```
Any reason not to inline the `filter_map()` into the `flat_map()` closure here?

djc commented on src/bin/cargo-flamegraph.rs:
```diff
@@ -316,11 +316,53 @@ fn workload(
+        .exec()
+        .expect("failed to access crate metadata");
+
+    let bin_targets: Vec<String> = metadata
```
I think we can chain this stuff together?

djc commented:
Sorry, I generated a bunch of conflicts here due to reformatting with the default line length. You should be able to just keep all your changes and reformat after merge/rebase.

ede1998 commented:
Ok, I implemented your suggestions or added a comment.

djc commented:
I'd like to have the reformatting applied in the earlier commits. This should be easy to automate with `git rebase -x "cargo fmt && git commit --amend -a"` or something like that (but if it's too much pain for you I'd be happy to do it myself).

ede1998 commented:
> I'd like to have the reformatting applied in the earlier commits.

Ok I misunderstood your previous comment then. I fixed the formatting again and implemented the new suggestions.

djc commented:
Thanks for all your work on this, this is a great improvement!
